core/src/main/java/com/scalar/db/api/ScanBuilder.java

```diff
@@ -321,11 +355,216 @@ public Scan build() {
+      return new BuildableScanAllWithOngoingWhereOr(this);
+    }
+
+    @Override
+    public BuildableScanAllWithOngoingWhereOr or(AndConditionSet andConditionSet) {
+      checkNotNull(super.condition);
+      checkNotNull(andConditionSet);
+      conjunctions.add(Scan.Conjunction.of(super.condition));
+      conjunctions.add(Scan.Conjunction.of(andConditionSet.getConditions()));
+      return new BuildableScanAllWithOngoingWhereOr(this);
+    }
+  }
+
+  public static class BuildableScanAllWithOngoingWhereAnd extends BuildableScanAllWithWhere
+      implements And<BuildableScanAllWithOngoingWhereAnd> {
+
+    private BuildableScanAllWithOngoingWhereAnd(BuildableScanAllWithOngoingWhere buildable) {
+      super(buildable);
+    }
+
+    private BuildableScanAllWithOngoingWhereAnd(
+        String namespaceName, String tableName, OrConditionSet orConditionSet) {
+      super(namespaceName, tableName);
+      disjunctions.add(orConditionSet.getConditions());
+    }
+
+    @Override
+    public BuildableScanAllWithOngoingWhereAnd and(ConditionalExpression condition) {
+      checkNotNull(condition);
+      disjunctions.add(ImmutableSet.of(condition));
+      return this;
+    }
+
+    @Override
+    public BuildableScanAllWithOngoingWhereAnd and(OrConditionSet orConditionSet) {
+      checkNotNull(orConditionSet);
+      disjunctions.add(orConditionSet.getConditions());
+      return this;
+    }
+  }
+
+  public static class BuildableScanAllWithOngoingWhereOr extends BuildableScanAllWithWhere
+      implements Or<BuildableScanAllWithOngoingWhereOr> {
+
+    private BuildableScanAllWithOngoingWhereOr(BuildableScanAllWithOngoingWhere buildable) {
+      super(buildable);
+    }
+
+    private BuildableScanAllWithOngoingWhereOr(
+        String namespaceName, String tableName, AndConditionSet andConditionSet) {
+      super(namespaceName, tableName);
+      conjunctions.add(Scan.Conjunction.of(andConditionSet.getConditions()));
+    }
+
+    @Override
+    public BuildableScanAllWithOngoingWhereOr or(ConditionalExpression condition) {
+      checkNotNull(condition);
+      conjunctions.add(Scan.Conjunction.of(condition));
+      return this;
+    }
+
+    @Override
+    public BuildableScanAllWithOngoingWhereOr or(AndConditionSet andConditionSet) {
+      checkNotNull(andConditionSet);
+      conjunctions.add(Scan.Conjunction.of(andConditionSet.getConditions()));
+      return this;
+    }
+  }
+
+  public static class BuildableScanAllWithWhere
+      implements Consistency<BuildableScanAllWithWhere>,
+          Projection<BuildableScanAllWithWhere>,
+          Ordering<BuildableScanAllWithWhere>,
+          Limit<BuildableScanAllWithWhere> {
+
+    protected final String namespaceName;
+    protected final String tableName;
+    @Nullable protected final ConditionalExpression condition;
+    protected final Set<Scan.Conjunction> conjunctions = new HashSet<>();
+    protected final List<Set<ConditionalExpression>> disjunctions = new ArrayList<>();
```

Comment:
Sorry for this delayed question. Why the types of `conjunctions` and `disjunctions` are different (`disjunctions` needs an order? `disjunctions` should be an internal class as well as `conjunctions`?)

Reply:
Thanks for the question. Actually, disjunctions don't need the order information, but `Sets.cartesianProduct()`, which is used for converting a CNF, requires a list.

Reply:
I see. If the reason of having different types for `conjunctions` and `disjunctions` is only for `Sets.cartesianProduct()`, we can write like this?

```
  Sets.cartesianProduct(
      disjunctions.stream().map(Disjunction::getConditions).collect(Collectors.toList()))
      .stream()
      .map(Conjunction::of)
      .collect(Collectors.toSet()));
```

It would be great if the types of `conjunctions` and `disjunctions` are the same-level abstraction because I guess the readers of this code expect that.

Reply:
Yeah, that's better for readers. Thank you. I think we can do just like the following. BTW, I'm not going to introduce `Scan.Disjunction` since the CNF and disjunctions are only used in the `ScanBuilder` internally at least at this moment. Even so, is it weird and should we have it?

```java
      scan.withConjunctions(
          Sets.cartesianProduct(new ArrayList<>(disjunctions)).stream()
              .map(Scan.Conjunction::of)
              .collect(Collectors.toSet()));
```

Reply:
> BTW, I'm not going to introduce Scan.Disjunction since the CNF and disjunctions are only used in the ScanBuilder internally at least at this moment. Even so, is it weird and should we have it?

@komamitsu Or, how about using `Set<Set<ConditionalExpression>>` for conjunctions when building and converting them at the last, similar to the disjunctions? What do you think?

Reply:
> how about using Set<Set<ConditionalExpression>> for conjunctions

It means like this?

```
    protected final Set<Set<ConditionalExpression>> conjunctions = new HashSet<>();
    protected final Set<Set<ConditionalExpression>> disjunctions = new HashSet<>();
```

If so, completely agreed.

Reply:
Yeah, that's what I meant. Fixed in a64862d. PTAL when you get a chance.